### PR TITLE
JDK-8272945: Use snippets in java.compiler documentation

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review a patch to use snippets in the `java.compiler` documentation, instead of a mix of raw HTML and/or `{@code ...}`.  The change is just about the presentation of the code fragments; there are no changes to the normative specifications for the module.

One of the examples went to extraordinary lengths to include the character sequence `*/` within the example. That example has been replaced by an external snippet in a separate source file, which does not have any restriction on the use of `*/`. However, it does require that the file be excluded from standard compilation, and that is done by setting `EXCLUDES`, once for the "interim" compiler, and once again for the "product" compiler.    Going forward, a better solution might be to modify the `SetupJavaCompilation` macro to ignore all directories whose name is not a valid Java identifier (or, if easier, contains a `-`, such as `doc-files` or `snippet-files`.)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272945](https://bugs.openjdk.java.net/browse/JDK-8272945): Use snippets in java.compiler documentation

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to 27467bf25cfba17f496d04bacf5400bb19dd6348
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6686/head:pull/6686` \
`$ git checkout pull/6686`

Update a local copy of the PR: \
`$ git checkout pull/6686` \
`$ git pull https://git.openjdk.java.net/jdk pull/6686/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6686`

View PR using the GUI difftool: \
`$ git pr show -t 6686`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6686.diff">https://git.openjdk.java.net/jdk/pull/6686.diff</a>

</details>

### CODE REVIEW

File: src/java.compiler/share/classes/javax/tools/snippet-files/JavaSourceFromString.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        super(URI.create("string:///" + name.replace('.','/') + Kind.SOURCE.extension), // @link substring="URI.create" target="URI#create(String)"
+              Kind.SOURCE);
+        this.code = code;
+    }
+
+    @Override
+    public CharSequence getCharContent(boolean ignoreEncodingErrors) {
+        return code;
+    }
+}
```
Comment (magicus) [inline src/java.compiler/share/classes/javax/tools/snippet-files/JavaSourceFromString.java:29]: Missing newline..?

File: make/modules/java.compiler/Java.gmk [hunk 2]
```diff
@@ -25,3 +25,6 @@
 
 DOCLINT += -Xdoclint:all/protected \
     '-Xdoclint/package:java.*,javax.*'
+
+EXCLUDES += \
+    javax/tools/snippet-files \
```
Comment (magicus) [inline make/modules/java.compiler/Java.gmk:30]: You can put this just on a single line :-). 

And I'm frankly not sure if make is happy about having a trailing backslash but no additional line...

File: src/java.compiler/share/classes/javax/tools/JavaCompiler.java [hunk 3]
```diff
@@ -181,32 +186,7 @@
...
- *                     Kind.SOURCE);
- *               this.code = code;
- *           }
- *
- *           {@code @}Override
- *           public CharSequence getCharContent(boolean ignoreEncodingErrors) {
- *               return code;
- *           }
- *       }</pre>
+ *     {@snippet id=fileObject class=JavaSourceFromString }
```
Comment (AlanBateman) [inline src/java.compiler/share/classes/javax/tools/JavaCompiler.java:189]: JEP 413 uses the "file" attribute for external snippets, so using "class" here is new (to me anyway). Maybe the JEP should include an example that uses it.

Is this the first use of an external snippet in the src tree? I suspect files in the snippet-files directory will need a copyright header (downstream buildes will report issues on this). It's not clear to me how that works if @replace region replacement="" isn't the first line of the external file.

File: make/modules/java.compiler/Java.gmk (hunk 2 above)
Comment (erikj79) [reply make/modules/java.compiler/Java.gmk:30]: As long as the next line is empty, it works, but it's not a good idea. That's why we came up with the terminating # in our 1 element per line lists. In this case, I would prefer a single line assignment without any backslashes.

File: src/java.compiler/share/classes/javax/tools/snippet-files/JavaSourceFromString.java (hunk 1 above)
Comment (jonathan-gibbons) [reply src/java.compiler/share/classes/javax/tools/snippet-files/JavaSourceFromString.java:29]: Oops

File: make/modules/java.compiler/Java.gmk (hunk 2 above)
Comment (jonathan-gibbons) [reply make/modules/java.compiler/Java.gmk:30]: Fixed to single line

**Review by erikj79 — APPROVED** [review]
Build changes look ok.

**Review by AlanBateman — APPROVED** [review]

File: src/java.compiler/share/classes/javax/tools/JavaCompiler.java (hunk 3 above)
Comment (jonathan-gibbons) [reply src/java.compiler/share/classes/javax/tools/JavaCompiler.java:189]: Added a legal header in the external snippet file and tested with a full `:tier1` build, including `validate-source`.

**Pull request closed without merging**